Add content cards documentation for Swift SDK (#4651)
braze-inc/braze-docs: _docs/_developer_guide/platform_integration_guides/swift/content_cards/integration.md

---
nav_title: Integration
article_title: Content Card View Controller Integration for iOS
platform: Swift
page_order: 1
description: "This article covers the integration steps, data models, and card-specific properties available for your iOS application."
channel:
  - content cards

---

# Content Card integration

## Content Cards data model

The Content Cards data model is available in the `BrazeKit` module of the iOS Swift SDK.

### Getting the data

To access the Content Cards data model, you can call `contentCards.cards` on your `braze` instance.

{% tabs %}
{% tab swift %}
```swift
let cards: [Braze.ContentCard] = AppDelegate.braze?.contentCards.cards
```

Additionally, you can also maintain a subscription to observe for changes in your content cards. You can do so in one of two ways: 1) Maintaining a cancellable, 2) Maintaining an `AsyncStream`.

{% subtabs %}
{% subtab CANCELLABLE %}
```swift
// This subscription is maintained through Braze cancellable, which will continue to observe for changes until the subscription is cancelled.

> let cancellable = AppDelegate.braze?.contentCards.subscribeToUpdates { [weak self] contentCards in

Reviewer: I think we can add wording [like in our public API](https://github.com/braze-inc/swift-sdk/blob/9a2acf4d02c81c0ebe38ef7ea9a536efbacc3a0d/Sources/BrazeKit/Api/ContentCards/ContentCardsApi.swift#L108) that you need a strong reference to this cancellable to keep it going

```

Additionally, if you wish to maintain a subscription to your content cards, you can call `subscribeToUpdates`:

```objc
// This subscription is maintained through Braze cancellable, which will continue to observe for changes until the subscription is cancelled.
BRZCancellable *cancellable = [self.braze.contentCards subscribeToUpdates:^(NSArray<BRZContentCardRaw *> *contentCards) {
  // Implement your completion handler to respond to updates in `contentCards`.
}];
```
{% endtab %}
{% endtabs %}

## Content Card model

Braze offers five Content Card types: banner, captioned image, classic, classic image, and control. Each type is an implementation of the `Braze.ContentCard` type. Note that BrazeKit offers an alternative [`ContentCardRaw`](https://braze-inc.github.io/braze-swift-sdk/documentation/brazekit/braze/contentcardraw) class for Objective-C compatibility.

For a full list of Content Card properties, as well as details about using Content Cards, refer to the [`ContentCard` class documentation](https://braze-inc.github.io/braze-swift-sdk/documentation/brazekit/braze/contentcard).

## Card methods

> Each card can be initialized with a `Context` object, which contains various methods for managing your card's state. Call these methods when you want to modify the corresponding state property on a particular card object.

Reviewer 2: ```suggestion
Each card is initialized with a `Context` object, which contains various methods for managing your card's state. Call these methods when you want to modify the corresponding state property on a particular card object.
```
Content cards created by the SDK always have a `context`.

> |Method|Description|
> |---|---|
> | `logImpression` | Log the content card impression event. |
> | `logClick` | Log the content card click event. |
> | `processClickAction` | Process a given [`ClickAction`](https://braze-inc.github.io/braze-swift-sdk/documentation/brazekit/braze/contentcard/clickaction) input. |
> | `logDismissed` | Log the content card dismissed event. |
> | `logError` | Log an error related to the content card. |
> | `loadImage` | Load a given content card image from a URL. This method can be nil when the content card does not have an image. |
> {: .reset-td-br-1 .reset-td-br-2}

Reviewer 2: Adds examples on how to call the methods.

```suggestion
| Method                               | Description                                                                                                                              |
|--------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------|
| `card.context?.logImpression()`      | Log the content card impression event.                                                                                                   |
| `card.context?.logClick()`           | Log the content card click event.                                                                                                        |
| `card.context?.processClickAction()` | Process a given [`ClickAction`](https://braze-inc.github.io/braze-swift-sdk/documentation/brazekit/braze/contentcard/clickaction) input. |
| `card.context?.logDismissed()`       | Log the content card dismissed event.                                                                                                    |
| `card.context?.logError()`           | Log an error related to the content card.                                                                                                |
| `card.context?.loadImage()`          | Load a given content card image from a URL. This method can be nil when the content card does not have an image.                         |
{: .reset-td-br-1 .reset-td-br-2}
```

btw, I use the following extension in VSCode to prettify the tables:
```
Name: Markdown Table Prettifier
Id: darkriszty.markdown-table-prettify
Description: Transforms markdown tables to be more readable.
Version: 3.6.0
Publisher: Krisztian Daroczi
VS Marketplace Link: https://marketplace.visualstudio.com/items?itemName=darkriszty.markdown-table-prettify
```

Author: That's a good tip, thanks! It feels inefficient to keep opening a preview to double-check if everything was formatted correctly 😅

| `card.context?.logError()`           | Log an error related to the content card.                                                                                                |
| `card.context?.loadImage()`          | Load a given content card image from a URL. This method can be nil when the content card does not have an image.                         |
{: .reset-td-br-1 .reset-td-br-2}

For more details, refer to the [`Context` class documentation](https://braze-inc.github.io/braze-swift-sdk/documentation/brazekit/braze/contentcardraw/context-swift.class)

## Content Cards UI integration

Content Cards UI can be integrated from the `BrazeUI` library of the Swift SDK. This library provides two view controller contexts: navigation or modal. If you wish to intercept and react to the Content Card UI lifecycle, implement [`BrazeContentCardUIViewControllerDelegate`](https://braze-inc.github.io/braze-swift-sdk/documentation/brazeui/brazecontentcarduiviewcontrollerdelegate) as the delegate for your `BrazeContentCardUI.ViewController`.

> ### Navigation context

Reviewer 3: I'm not sure I understand the difference between a navigation or modal context. Unsure if this is something that is known by all of our intended users, but I think it would be worthwhile to provide more context about these contexts (haha, forgive me) to accommodate developer users of all experience levels.

Reviewer: Maybe we can add a GIF to help highlight the differences, but it is an iOS concept

Modal comes from bottom with a separate element:
https://i.stack.imgur.com/2XEO4.gif

Navigation goes from side:
https://i.stack.imgur.com/1yNaF.gif

Reviewer 3: Thanks for the clarification! The link helps a lot.
